### PR TITLE
TRUNK-5807, Exception thrown in console on saving Patient in PostgreSQL

### PR DESCRIPTION
**Pull request by Am-Coder**
## Description of what I changed

Problem is presence of orderby does not affect aggregate function in mysql but in postgresql it does. So whereever you are counting you need to remove order clause in criteria. The SQL query that gets executed is :

select count(*) as y0_ from visit this_ where this_.patient_id=? and not exists (select 1 from encounter where this_.visit_id=visit_id) and this_.voided=? order by this_.date_started desc, this_.visit_id desc

So had to remove order by clause and there is no direct way to get rid of it. I referred https://stackoverflow.com/questions/2370221/remove-dynamically-an-ordering-to-the-result-set-in-org-hibernate-criteria  
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5807

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateEncounterDAO.java [hunk 1]
```diff
@@ -588,12 +590,26 @@ public Integer getEncountersByVisitsAndPatientCount(Patient patient, boolean inc
 		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Visit.class);
 		addEmptyVisitsByPatientCriteria(criteria, patient, includeVoided, query);
 		
+		//A hack to remove the order_by clauses as they cause error in PostgreSQL
+		Iterator<CriteriaImpl.OrderEntry> orderIter = ((CriteriaImpl) criteria).iterateOrderings();
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateEncounterDAO.java:594]: Why use a hack?

Comment (Am-Coder) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateEncounterDAO.java:594]: There is no direct way to remove the order clause from criteria once it gets added( [docs](https://docs.jboss.org/hibernate/orm/3.3/reference/en/html/querycriteria.html) ).  I found this solution [here](https://stackoverflow.com/questions/2370221/remove-dynamically-an-ordering-to-the-result-set-in-org-hibernate-criteria).

**Comment by gitcliff** [comment]
hello @dkayiwa i think we can merge this .....
LGTM

**Comment by dkayiwa** [comment]
@Am-Coder do you know why this error does not happen in MySQL?

**Comment by Am-Coder** [comment]
> @Am-Coder do you know why this error does not happen in MySQL?

The query that gets executed is this : 

`select count(*) as y0_ from visit this_ where this_.patient_id=? and not exists (select 1 from encounter where this_.visit_id=visit_id) and this_.voided=? order by this_.date_started desc, this_.visit_id desc
`
Presence of order by clause when there is only aggregate function in select section does not cause error on MySQL but on PostgreSQL it does. I debugged the query getting executed and this was the cause of error.

**Comment by dkayiwa** [comment]
Do you know why it does not cause an error on MySQL?

**Comment by ibacher** [comment]
@dkayiwa If I had to venture a guess, it's because the MySQL parser is a little more lenient in what it allows.

In any case, from looking at the `explain` output, it's clear that the MySQL optimizer doesn't do anything with the `order by` clause if using an aggregate function:

```sql
explain select distinct obs.uuid
from obs
order by obs.date_created;
```

```
+--+-----------+-----+----+--------------+----+-------+----+----+--------------+
|id|select_type|table|type|possible_keys |key |key_len|ref |rows|Extra         |
+--+-----------+-----+----+--------------+----+-------+----+----+--------------+
|1 |SIMPLE     |obs  |ALL |obs_uuid_index|NULL|NULL   |NULL|4   |Using filesort|
+--+-----------+-----+----+--------------+----+-------+----+----+--------------+
```

```sql
explain select count(distinct obs.uuid)
from obs
order by obs.date_created;
```

```
+--+-----------+-----+----+--------------+----+-------+----+----+-----+
|id|select_type|table|type|possible_keys |key |key_len|ref |rows|Extra|
+--+-----------+-----+----+--------------+----+-------+----+----+-----+
|1 |SIMPLE     |obs  |ALL |obs_uuid_index|NULL|NULL   |NULL|4   |NULL |
+--+-----------+-----+----+--------------+----+-------+----+----+-----+
```

This is probably to make it easier to play around with things, since, e.g., the below is a perfectly valid query:

```sql
select count(distinct obs.uuid)
from obs
group by year(obs.date_created)
order by year(obs.date_created);
```

And if I tweak it by commenting out the `group by` clause, we get a valid query in MySQL that ignores the `order by` clause, but a broken query in PostgreSQL.

**Comment by dkayiwa** [comment]
Oh i see!!! Thank you so much @ibacher 👍

**Pull request merged**